server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -1135,13 +1109,36 @@ private static IndexOptions parseIndexOptions(String fieldName, Object propNode)
      * @return the custom kNN vectors format that is configured for this field or
      * {@code null} if the default format should be used.
      */
-    public KnnVectorsFormat getKnnVectorsFormatForField() {
+    public KnnVectorsFormat getKnnVectorsFormatForField(KnnVectorsFormat defaultFormat) {
+        KnnVectorsFormat format;
         if (indexOptions == null) {
-            return null; // use default format
+            format = defaultFormat;
         } else {
             HnswIndexOptions hnswIndexOptions = (HnswIndexOptions) indexOptions;
-            return new Lucene95HnswVectorsFormat(hnswIndexOptions.m, hnswIndexOptions.efConstruction);
+            format = new Lucene95HnswVectorsFormat(hnswIndexOptions.m, hnswIndexOptions.efConstruction);
         }
+        // It's legal to reuse the same format name as this is the same on-disk format.
+        return new KnnVectorsFormat(format.getName()) {
+            @Override
+            public KnnVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException {
+                return format.fieldsWriter(state);
+            }
+
+            @Override
+            public KnnVectorsReader fieldsReader(SegmentReadState state) throws IOException {
+                return format.fieldsReader(state);
+            }
+
+            @Override
+            public int getMaxDimensions(String fieldName) {
+                return MAX_DIMS_COUNT;
```

Comment:
In 8.11, we want to increase MAX_DIMS_COUNT to 4096.  Will we need to create a new KnnVectorsFormat for that?

Reply:
No need to create a new format, we can simply bump the constant from 2048 to 4096. I think we should allow existing indexes to have more than 2048 dimensions too indeed.